src/main/java/com/vlkan/pubsub/PubsubClient.java

```diff
@@ -0,0 +1,274 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permits and
+ * limitations under the License.
+ */
+
+package com.vlkan.pubsub;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.vlkan.pubsub.model.*;
+import com.vlkan.pubsub.util.MicrometerHelpers;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Timer;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+
+import javax.annotation.Nullable;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Objects;
+import java.util.WeakHashMap;
+import java.util.function.Supplier;
+
+public class PubsubClient {
+
+    public static final ObjectMapper DEFAULT_OBJECT_MAPPER = new ObjectMapper()
+            // To allow backward-compatible future enhancements in the
+            // protocol, disable failure on unknown properties.
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+
+    private final PubsubClientConfig config;
+
+    private final ObjectMapper objectMapper;
+
+    private final PubsubAccessTokenCache accessTokenCache;
+
+    private final HttpClient httpClient;
+
+    private final MeterRegistry meterRegistry;
+
+    private final Map<String, Timer> timerByRequestUrl = new WeakHashMap<>();
+
+    private final Map<String, Counter> counterByRequestUrl = new WeakHashMap<>();
+
+    private PubsubClient(Builder builder) {
+        this.config = builder.config;
+        this.objectMapper = builder.objectMapper;
+        this.accessTokenCache = builder.accessTokenCache;
+        this.httpClient = builder.httpClient;
+        this.meterRegistry = builder.meterRegistry;
+    }
+
+    Mono<PubsubPullResponse> pull(String projectName, String subscriptionName, PubsubPullRequest pullRequest) {
+        String requestUrl = String.format(
+                "%s/v1/projects/%s/subscriptions/%s:pull",
+                config.getBaseUrl(), projectName, subscriptionName);
+        Supplier<String[]> tagSupplier = () -> new String[]{
+                "operation", "pull",
+                "projectName", projectName,
+                "subscriptionName", subscriptionName
+        };
+        return executeRequest(requestUrl, pullRequest, PubsubPullResponse.class, config.getPullTimeout())
+                .transform(mono -> MicrometerHelpers.measureLatency(
```

Comment:
is there a reason you did not want to use Reactor's [built-in MicroMeter support](https://projectreactor.io/docs/core/milestone/reference/#_publisher_metrics)?

Reply:
Yes, there is a very valid justification: I didn't even know that such a thing existed at all! Nevertheless, I've checked the `metrics()` directive and noticed that it doesn't support custom tags, which is crucial in my case. (Am I missing something?)

In the meantime, I have noticed that accesses to `Map<String, Timer>` and `Map<String, Counter>` maps in `MicrometerHelpers` are not synchronized. I will wrap those `WeakHashMap`s with `Collections.synchronizedMap()`. That is valid, right? (Or shall I rather `synchronize (map) { ... }` in `MicrometerHelpers`?)

Reply:
> it doesn't support custom tags

It does, see https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#tag-java.lang.String-java.lang.String-

Reply:
Ok, you got me there. There is one more justification in my sleeve: I only add (i.e., allocate entries for) tags if there doesn't exist a previously created meter for the given `requestUrl`. That is, if there is already an associated meter in `{timer,counter}ByRequestUrl` caches, I only waste a `Supplier<String[]>` allocation. Whereas, I guess, `Mono#tag()` will allocate those entries at every call. WDYT?

Reply:
this could be the case, although I would not bother too much since it is a micro optimization here.

I would still prefer to use what Reactor provides and report any perf issue discovered (if it affects you).

P.S. we also have some exciting ideas how to optimize it, stay tuned :)